File: feature/base/src/main/java/net/twisterrob/android/view/CheckableButtonManager.java
package net.twisterrob.android.view;

import java.util.*;

import android.widget.CompoundButton;

public class CheckableButtonManager implements CompoundButton.OnCheckedChangeListener {
	private final List<CompoundButton> buttons = new LinkedList<>();
	private CompoundButton checkedButton = null;
	private static final boolean safeMode = false;
	private CompoundButton.OnCheckedChangeListener listener;

	public void addButton(CompoundButton rb) {
		if (rb.isChecked()) {
			if (checkedButton == null) {
				checkedButton = rb;
			} else {
				rb.setChecked(false);
			}
		}
		rb.setOnCheckedChangeListener(this);
		buttons.add(rb);
	}

	public void onCheckedChanged(CompoundButton newlyCheckedButton, boolean isChecked) {
		if (isChecked) {
			disable(newlyCheckedButton);
			checkedButton = newlyCheckedButton;
			fireCheckedChange();
		}
	}

	public void setOnCheckedChangeListener(CompoundButton.OnCheckedChangeListener listener) {
		this.listener = listener;
	}

	private void fireCheckedChange() {
		if (listener != null) {
			listener.onCheckedChanged(checkedButton, true);
		}
	}

	private void disable(CompoundButton newlyCheckedButton) {
		if (safeMode) {
			for (CompoundButton cb : buttons) {
				if (!newlyCheckedButton.equals(cb)) {
					cb.setChecked(false);
				}
			}
		} else {
			if (checkedButton != null) {
				checkedButton.setChecked(false);
			}
		}
	}

	public CompoundButton getChecked() {
		return checkedButton;
	}
}

File: feature/base/src/main/kotlin/net/twisterrob/android/view/CheckableButtonManager.kt
package net.twisterrob.android.view

import android.widget.CompoundButton
import java.util.LinkedList

class CheckableButtonManager : CompoundButton.OnCheckedChangeListener {

	private val buttons = LinkedList<CompoundButton>()
	private var listener: CompoundButton.OnCheckedChangeListener? = null
	var checked: CompoundButton? = null; private set

	fun addButton(rb: CompoundButton) {
		updateChecked(rb)
		rb.setOnCheckedChangeListener(this)
		buttons.add(rb)
	}

	private fun updateChecked(rb: CompoundButton) {
		if (rb.isChecked) {
			if (checked == null) {
				checked = rb
			} else {
				rb.isChecked = false
			}
		}
	}

	override fun onCheckedChanged(newlyCheckedButton: CompoundButton, isChecked: Boolean) {
		if (isChecked) {
			disable(newlyCheckedButton)
			checked = newlyCheckedButton
			fireCheckedChange(newlyCheckedButton)
		}
	}

	fun setOnCheckedChangeListener(listener: CompoundButton.OnCheckedChangeListener) {
		this.listener = listener
	}

	private fun fireCheckedChange(checked: CompoundButton) {
		listener?.onCheckedChanged(checked, true)
	}

	private fun disable(newlyCheckedButton: CompoundButton) {
		if (safeMode) {
			buttons.filterNot { it == newlyCheckedButton }.forEach { it.isChecked = false }
		} else {
			checked?.isChecked = false
		}
	}

	companion object {

		private const val safeMode = false
	}
}
